horizons/util/platform.py

```diff
@@ -31,21 +30,86 @@ def get_home_directory():
 	Returns the home directory of the user running UH.
 	"""
 	if platform.system() != "Windows":
-		return PurePath(os.path.expanduser('~'))
+		return os.path.expanduser('~')
 	else:
 		import ctypes.wintypes
 		buf = ctypes.create_unicode_buffer(ctypes.wintypes.MAX_PATH)
 		# get the My Documents folder into buf.value
 		ctypes.windll.shell32.SHGetFolderPathW(0, CSIDL_PERSONAL, 0, 0, buf)
-		return PurePath(buf.value)
+		return buf.value
 
 
-def get_user_game_directory():
+# define here as string to reduce chance of typos
+UH = "unknown-horizons"
+
+
+def get_old_user_game_directory():
 	"""
-	Returns the directory where we store game-related data, such as savegames.
+	Returns the old directory where for game-related data.
+	This is used when migrating from the old structure to the new
 	"""
+	user_dir = os.environ.get("UH_USER_DIR")
+	if user_dir: # not None or empty string
+		return user_dir
+
 	home_directory = get_home_directory()
 	if platform.system() != "Windows":
-		return home_directory.joinpath('.unknown-horizons')
+		return os.path.join(home_directory, "." + UH)
 	else:
-		return home_directory.joinpath('My Games', 'unknown-horizons')
+		return os.path.join(home_directory, 'My Games', UH)
+
+
+def _try_directories(paths, default=None):
+	# takes a list of lists of strings as first argument
+	# returns the joined path for the first list where all strings exist and are nonempty
+	# or the second argument there is no such list
+	# (lists coul also be tuples of course)
+	for parts in paths:
+		if all(parts):
+			return os.path.join(parts)
+	return default
+
+
+def get_user_game_directories():
+	"""
+	Returns a triplet of directories for game-related data.
+	The first value is the directory for configuration files (settings.xml)
+	The second is the directory for game data (saves, maps, screenshots, logs)
+	The third is the directory for cache (atlas-metadata.cache, yamldata.cache)
+	"""
+	home_directory = get_home_directory()
+
+	# these are the default directories
+	if platform.system() == "Windows":
+		user_dir = os.path.join(home_directory, 'My Games', UH)
```

Comment:
In case of Windows the user_dir is set, otherwise it isn't. Is this wanted?

Reply:
user_dir is only a helper variable in this function. It's scope is this function and it is not returned. It is just to make the next 3 assignments a bit shorter.

Reply:
I just found it a bit confusing, that there was a difference between Win and non-Win.

Reply:
There is always a difference between windows and non-windows because windows doesn't have `.cache`, `.local/share` and `.config` directories. That's why all uh user data is sill in the same directory (I haven't found a much better place to put it). It made sense to me to put this common directory in a helper variable.